mlflow/store/artifact_repository_registry.py

```diff
@@ -0,0 +1,103 @@
+from mlflow.store.sftp_artifact_repo import SFTPArtifactRepository
+from mlflow.store.dbfs_artifact_repo import DbfsArtifactRepository
+from mlflow.store.s3_artifact_repo import S3ArtifactRepository
+from mlflow.store.local_artifact_repo import LocalArtifactRepository
+from mlflow.store.rest_store import RestStore
+
+
+class ArtifactRepositoryRegistry:
+    """Scheme-based registry for artifact repository implementations
+
+    This class allows the registration of a function or class to provide an implementation for a
+    given scheme of `artifact_uri` through the `register` method. Implementations declared though
+    the entrypoints `mlflow.artifact_repository` group can be automatically registered through the
+    `register_entrypoints` method.
+
+    When instantiating an artifact repository through the `get_artifact_repository` method, the
+    scheme of the artifact URI provided will be used to select which implementation to instantiate,
+    which will be called with same arguments passed to the `get_artifact_repository` method.
+    """
+
+    def __init__(self):
+        self._registry = {}
+
+    def register(self, scheme, repository):
+        """Register artifact repositories provided by other packages"""
+        self._registry[scheme] = repository
+
+    def register_entrypoints(self):
+        # Register artifact repositories provided by other packages
+        for entrypoint in entrypoints.get_group_all("mlflow.artifact_repository"):
+            try:
+                self.register(entrypoint.name, entrypoint.load())
+            except (AttributeError, ImportError) as exc:
+                warnings.warn(
+                    'Failure attempting to register artifact repository for scheme "{}": {}'.format(
+                        entrypoint.name, str(exc)
+                    ),
+                    stacklevel=2
+                )
+
+    def get_artifact_repository(self, artifact_uri, store=None):
+        """Get an artifact repository from the registry based on the scheme of artifact_uri
+
+        :param store_uri: The store URI. This URI is used to select which artifact repository
+                          implementation to instantiate and is passed to the
+                          constructor of the implementation.
+        :param store: Instance of `AbstractStore`. This is currently only used to get the host
+                      credentials when instantiating a DBFS-backed artifact repository.
+
+        :return: An instance of `mlflow.store.ArtifactRepository` that fulfills the artifact URI
+                 requirements.
+        """
+        scheme = urllib.parse.urlparse(artifact_uri).scheme
+        repository = self._registry.get(scheme)
+        if scheme == "dbfs" and repository is not None:
+            if not isinstance(store, RestStore):
+                raise MlflowException('`store` must be an instance of RestStore.')
+            return repository(artifact_uri, store.get_host_creds)
```

Comment:
As discussed in [this gist](https://gist.github.com/acroz/48a7b9455f1a054f6b14acf676d6df46), we have plans to deprecate this call-out of DBFS as a special case.